src/NuGet.Clients/VsExtension/VisualStudioAccountProvider.cs

```diff
@@ -0,0 +1,390 @@
+
+            ICredentials ret;
+            var posibleAccounts = new List<AccountAndTenant>();
+
+            //  Check to see if this is a VSO endpoint before we do anything else
+            var uriTenantId = await LookupTenant(uri, proxy, cancellationToken).ConfigureAwait(false);
+            if (string.IsNullOrWhiteSpace(uriTenantId))
+            {
+                //  we don't have a tenant ID so this cannot be a VSO endpoint
+                return null;
+            }
+
+            if (_accountManager == null)
+            {
+                // we know this is a VSO endpoint but we are unable to get the credentials so we should
+                // throw so that the other providers will not be called
+                throw new Exception("could not get the account manager, unable to check the keychain");
+            }
+
+            VSAccountProvider provider = null;
+            provider = (VSAccountProvider) await _accountManager
+                .GetAccountProviderAsync(VSAccountProvider.AccountProviderIdentifier).ConfigureAwait(false);
+
+            if (provider == null)
+            {
+                // we know this is a VSO endpoint but we are unable to get the credentials so we should
+                // throw so that the other providers will not be called
+                throw new Exception("could not get the account provider, unable to check the keychain");
+            }
+
+            //  Ask keychain for all accounts
+            var accounts = _accountManager.Store.GetAllAccounts();
+
+            //  Look through the accounts to see what ones have the VSO tenant in them (collected from
+            //  the LookupTenant() call at the top of the method).
+            foreach (var account in accounts)
+            {
+                var tenant = FindTenantInAccount(account, uriTenantId, provider);
+                if (tenant != null)
+                {
+                    posibleAccounts.Add(new AccountAndTenant(account, tenant));
+                }
+            }
+
+            if (posibleAccounts.Count == 1)
+            {
+                //  If we only have one posible account use it
+                ret = await GetTokenFromAccount(posibleAccounts[0], provider, nonInteractive, cancellationToken)
+                    .ConfigureAwait(false);
+                if (isRetry)
+                {
+                    var hasAccess = await AccountHasAccess(uri, proxy, ret, cancellationToken)
+                        .ConfigureAwait(false);
+                    if (!hasAccess)
+                    {
+                        // The account didn't have access and we are on a retry so the token didn't expire.
+                        // we either need to prompt the user for different creds or fail in this case
+                        ret = await PromptUserForAccount(uriTenantId, provider, nonInteractive, cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                }
```

Comment:
GetTokenFromAccount gets a session token and it will be a new token not the cached token from the previous try.  So in the case where is retry true and has access true we are refreshing the session token.